src/main/proto/firrtl.proto

```diff
@@ -0,0 +1,451 @@
+      Direction direction = 1;
+      // Required.
+      string id = 2;
+      // Required.
+      string memory_id = 3;
+      // Required.
+      Expression memory_index = 4;
+      // Required.
+      Expression expression = 5;
+    }
+
+    // Required.
+    oneof statement {
+      Wire wire = 1;
+      Register register = 2;
+      Memory memory = 3;
+      CMemory cmemory = 4;
+      Instance instance = 5;
+      Node node = 6;
+      When when = 7;
+      Stop stop = 8;
+      Printf printf = 10;
+      InlineVerilog inline_verilog = 11;
+      Assert assert = 12;
+      Cover cover = 13;
+      Skip skip = 14;
+      Connect connect = 15;
+      PartialConnect partial_connect = 16;
+      IsInvalid is_invalid = 17;
+      MemoryPort memory_port = 18;
+    }
+
+    SourceInfo source_info = 19;
+  }
+
+  // Using proto3 means that there is no has* method for primitives. This
+  // necesitates boxing width values because there are cases where a width of
+  // zero (the default value of uint32) is a valid width.
+  message Width {
+    // Required
+    uint32 value = 1;
+  }
+
+  message Type {
+    message UIntType {
+      Width width = 1;
+    }
+
+    message SIntType {
+      Width width = 1;
+    }
+
+    message ClockType {
+      // Empty.
+    }
+
+    message RecordType {
+      message Field {
+        // Required.
+        bool is_flipped = 1;
+        // Required.
+        string id = 2;
+        // Required.
+        Type type = 3;
+      }
+      repeated Field field = 1;
+    }
+
+    message VectorType {
+      // Required.
+      Type type = 1;
+      // Required.
+      uint32 size = 2;
+    }
+
+    // Required.
+    oneof type {
+      UIntType uint_type = 2;
+      SIntType sint_type = 3;
+      ClockType clock_type = 4;
+      RecordType record_type = 5;
+      VectorType vector_type = 6;
+    }
+  }
+
+  message Port {
+    enum Direction {
+      PORT_DIRECTION_UNKNOWN = 0;
+      PORT_DIRECTION_IN = 1;
+      PORT_DIRECTION_OUT = 2;
+    }
+
+    // Required.
+    string id = 1;
+    // Required.
+    Direction direction = 2;
+    // Required.
+    Type type = 3;
+  }
+
+  message Expression {
+    message Reference {
+      // Required.
+      string id = 1;
+    }
+
+    message IntegerLiteral {
+      // Base 10 value. May begin with a sign (+|-). Only zero can begin with a
+      // '0'.
+      // Required
+      string value = 1;
+    }
+
+    message UIntLiteral {
+      // Required.
+      IntegerLiteral value = 1;
+      Width width = 2;
+    }
+
+    message SIntLiteral {
+      // Required.
+      IntegerLiteral value = 1;
+      Width width = 2;
+    }
+
+    message ValidIf {
+      // Required.
+      Expression condition = 1;
+      // Required.
+      Expression value = 2;
+    }
+
+    message Mux {
+      // Required.
+      Expression condition = 1;
+      // Required.
+      Expression t_value = 2;
+      // Required.
+      Expression f_value = 3;
+    }
+
+    message ExtractBits {
```

Comment:
Not sure why this is separate from PrimOp?

Reply:
Fine by me. I am not sure what was behind that decision. Similarly Mux, SubIndex, SubAccess could be changed to.